openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1577 @@
+    template<typename MaskTreeT>
+    bool initMask(const SdfGridT &sdfGrid, const Grid<MaskTreeT> &mask, bool ignoreActiveTiles = false);
+
+    /// @brief Perform @a nIter iterations of the fast sweeping algorithm.
+    ///
+    /// @param nIter Number of iterations of the fast sweeping algorithm.
+    ///              Each iteration performs 2^3 = 8 individual sweeps.
+    ///
+    /// @param finalize If true the (possibly asymmetric) inside and outside values of the
+    ///                 resulting signed distance field are properly set. Unless you're
+    ///                 an expert this should remain true!
+    ///
+    /// @throw RuntimeError if sweepingVoxelCount() or boundaryVoxelCount() return zero.
+    ///        This might happen if none of the initialization methods above were called
+    ///        or if that initialization failed.
+    void sweep(int nIter = 1, bool finalize = true);
+
+    /// @brief Clears all the grids and counters so initializtion can be called again.
+    void clear();
+
+    /// @brief Return the number of voxels that will be solved for.
+    size_t sweepingVoxelCount() const { return mSweepingVoxelCount; }
+
+    /// @brief Return the number of voxels that defined the boundary condition.
+    size_t boundaryVoxelCount() const { return mBoundaryVoxelCount; }
+
+    /// @brief Return true if there are voxels and boundaries to solve for
+    bool isValid() const { return mSweepingVoxelCount > 0 && mBoundaryVoxelCount > 0; }
+
+private:
+
+    /// @brief Private method to prune the sweep mask and cache leaf origins.
+    void computeSweepMaskLeafOrigins();
+
+    // Private utility classes
+    template<typename>
+    struct MaskKernel;//   initialization to extend a SDF into a mask
+    template<typename>
+    struct InitExt;
+    struct InitSdf;
+    struct DilateKernel;// initialization to dilate a SDF
+    struct MinMaxKernel;
+    struct SweepingKernel;// performs the actual concurrent sparse fast sweeping
+
+    // Define the topology (i.e. stencil) of the neighboring grid points
+    static const Coord mOffset[6];// = {{-1,0,0},{1,0,0},{0,-1,0},{0,1,0},{0,0,-1},{0,0,1}};
+
+    // Private member data of FastSweeping
+    typename SdfGridT::Ptr mSdfGrid;
+    typename ExtGridT::Ptr mExtGrid;
+    SweepMaskTreeT mSweepMask; // mask tree containing all non-boundary active voxels
+    std::vector<Coord> mSweepMaskLeafOrigins; // cache of leaf node origins for mask tree
+    size_t mSweepingVoxelCount, mBoundaryVoxelCount;
+};// FastSweeping
+
+////////////////////////////////////////////////////////////////////////////////
+
+// Static member data initialization
+template <typename SdfGridT, typename ExtValueT>
+const Coord FastSweeping<SdfGridT, ExtValueT>::mOffset[6] = {{-1,0,0},{1,0,0},
+                                                             {0,-1,0},{0,1,0},
+                                                             {0,0,-1},{0,0,1}};
```

Comment:
Can you not use the offsets available in `openvdb/tools/Utils.h`

Reply:
not really - the ordering is different and I do a lot of index access like mOffset[i]. Changing the order would be a mess

Reply:
Oops, didn't realise the orders were different